coastguard/coastguard.md

```diff
@@ -0,0 +1,138 @@
+
+## Motivation
+With the advent of [multi cluster services api](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api) the need for restricting traffic across a cluster-set has become more apparent.
+
+When a service is imported into a cluster, the imported service and it's underlying pods are accessible from all pods in the importing cluster. 
+
+Limiting traffic to the service and it's underlying pods by other pods can't be done by semantic rules (namespaces, labels), which greatly limits the options for a cluster administrator wanting to limit and secure his cross-service traffic.
+## Requirements
+1. Connected Clusters without any network policies should keep the default behavior.
+2. Allow limiting traffic to/and from a service based on namespace.
+3. Allow limiting traffic to/and from the underlying pods of a service based on labels and/or namespace.
+4. When a cluster previously connected via submariner is removed, any auxiliary resources, rules, or artifacts maintained by the submariner network policy implementation must be properly cleaned up.
+
+## Proposal Architecture
+
+![image](https://user-images.githubusercontent.com/42589698/146993515-ba6bc995-8d3b-4642-b95f-dc56f4083502.png)
+
+Coastguard aims to implement multi-cluster network policies by managing single-cluster network policies in each cluster. 
+Our main challenge is tracking changes in deployments and pods in each cluster, and thus keeping the single-cluster network policies up to date. 
+Coastguard is composed of:
+  1.	Operator pattern = k8s controller + custom CRD. The custom crd will define the multi-cluster network policies.
+  2.	Controller will use embedded etcd. as etcd is a distributed key-value store, etcd is responsible for updating the changes from the local cluster, to the controller in the remote cluster.
+
+  3.	Controller tracks deployments using the Admiral API that federates resources to submariner data planes running in multiple clusters. Admiral api is an existing code component in submariner. 
+
+For security reasons, our technique uses the existing APIs and allows updates through them, so that an admin who controls the network policy of a cluster can only update changes in the CRD through the API and has no other authorization over the clusters. This model still has security implications that need to be explored and tested.
+
+The core idea here is: that each controller has a reconciliation loop, where it re-generates a local policy based on remote network policy. Each controller has an embedded etcd server which holds the ip address and labels of all pods across all clusters. The controller will generate a “local”/regular network policy and will apply it to k8s api.
```

Comment:
So is the etcd stretched across the cluster set? How is this etcd communicating? Who is responsible to allow it to communicate with other instances? How does it know the address of other etcd servers?